### PR TITLE
Add parser-level check for missing semicolons

### PR DESCRIPTION
**Pull request by kaicataldo**
Fixes https://github.com/adobe-community/issues/issues/11588

This adds an actual check to see if a semicolon is present at the end of rule declarations. This prevents a rule declaration from being multiline, but does not prevent declarations from being on the same line, provided they each terminate with a semicolon.

Most of this new method is copied from https://github.com/postcss/postcss/blob/master/lib/parser.es6#L167

### CODE REVIEW

File: src/parser/parse.js [hunk 1]
```diff
@@ -1,10 +1,108 @@
...
+    }
+
+    if (node.prop[0] === '_' || node.prop[0] === '*') {
+      node.raws.before += node.prop[0];
+      node.prop = node.prop.slice(1);
+    }
+    node.raws.between += this.spacesFromStart(tokens);
+    this.precheckMissedSemicolon(tokens);
+
+    for (let i = tokens.length - 1; i > 0; i--) {
```
Comment (weikinhuang) [inline src/parser/parse.js]: is it ok that you're modifying tokens further up?

Comment (kaicataldo) [inline src/parser/parse.js]: This is how the original parser does it. At the point that my new method `validateValue` executes, the remaining tokens that are left are only the value of the rule declaration. Interestingly enough, it seems like there isn't any validation that occurs for the value itself (except that multiple colons don't precede a semicolon, which is what I've removed and replaced here).

Comment (kaicataldo) [inline src/parser/parse.js]: This is also why I have to loop over the remaining tokens (that are assumed to be the value of the declaration) - it doesn't actually ever iterate over them...

**Comment by weikinhuang** [comment]
:+1:

**Comment by matt-oconnell** [comment]
:+1:

File: test/parser/index.js [hunk 2]
```diff
@@ -396,25 +423,47 @@ describe('parser', function() {
...
+        const styles = `.missing-var {
                                      prop: $not-defined;
                                    }`;
         const errMsg = 'Variable $not-defined is not defined. (2:38)';
 
         it('when parsed with parse', function(done) {
-          parse(undefinedVarStyles)
-          .catch(({ message }) => {
+          parse(styles)
+          .then(done.fail, ({ message }) => {
```
Comment (kaicataldo) [inline test/parser/index.js]: Realized this test would not fail when the parse does succeeds.

**Comment by kaicataldo** [comment]
Realized there is a MUCH better solution - sorry guys, mind re-reviewing in the morning?

**Comment by weikinhuang** [comment]
Nice!

**Comment by weikinhuang** [comment]
👍

**Comment by matt-oconnell** [comment]
👍

**Pull request merged**